File: flutter_app/lib/core/utils/birthday.dart
// ============================================
// 生日工具 (主人 2026-09-18 拍: 生日提醒)
//
// 支持阳历 / **农历** 两种历法 → 算「距下次生日还有几天」/「下个生日是几月几号」
// 农历转换用 `lunar` 包 (6tail, MIT): Lunar.fromYmd(农历年, 月, 日).getSolar()
//
// 边界 (重要):
//   - 月 / 日 任一为空 → 算不出来, 返回 null (客户只填了年份也要能跑)
//   - 农历: 闰月不参与 (生日按普通月算, 正数月); 若某年该月是闰月 → 取正月初一那天
//   - 2/29 (阳历) 在平年 → 按 3/1 算? 不, 按「2 月最后一天」= 2/28 (主人可用, 不报错)
//   - 只做「未来 0-370 天」范围, 不回溯历史
// ============================================

import 'package:lunar/lunar.dart';

class BirthdayInfo {
  /// 距下次生日还有几天 (0 = 今天就是生日)
  final int daysUntil;
  /// 下次生日的阳历日期 (yyyy-MM-dd)
  final DateTime nextSolarDate;
  /// 生日文案 (例: '八月十五 (农历)' / '3月8日' / '只知道年份 1965')
  final String label;

  const BirthdayInfo({
    required this.daysUntil,
    required this.nextSolarDate,
    required this.label,
  });

  String get countdownLabel {
    if (daysUntil == 0) return '今天生日 🎂';
    if (daysUntil == 1) return '明天生日';
    return '$daysUntil 天后生日';
  }
}

/// 农历月日 → 文案 (数字转中文, 中老年看得懂)
String _lunarChinese(int month, int day) {
  const digits = ['', '一', '二', '三', '四', '五', '六', '七', '八', '九', '十'];
  String m() {
    if (month == 1) return '正月';
    if (month <= 10) return '${digits[month]}月';
    return month == 11 ? '冬月' : '腊月';
  }

  String d() {
    if (day <= 10) return '初${digits[day] == '十' ? '十' : digits[day]}';
    if (day < 20) return '十${digits[day - 10]}';
    if (day == 20) return '二十';
    if (day < 30) return '廿${digits[day - 20]}';
    return day == 30 ? '三十' : '廿${digits[day - 20]}';
  }

  return '${m()}${d()}';
}

/// 生日文案 (列表/详情显示用)
String birthdayLabel({
  required int? month,
  required int? day,
  String calendar = 'solar',
  int? year,
}) {
  if (month == null || day == null) {
    if (year != null) return '只知道年份 $year';
    return '未填';
  }
  if (calendar == 'lunar') {
    return '${_lunarChinese(month, day)} (农历)';
  }
  return '$month月$day日';
}

/// 下次生日的阳历日期 (含今天); 算不出来返回 null
DateTime? nextBirthdayDate({
  required int? month,
  required int? day,
  String calendar = 'solar',
  DateTime? from,
}) {
  if (month == null || day == null) return null;
  final today = from ?? DateTime.now();
  final base = DateTime(today.year, today.month, today.day);

  if (calendar == 'lunar') {
    // 农历: 试当年 + 下一年, 取第一个 >= 今天
    for (final y in [base.year, base.year + 1]) {
      try {
        final solar = Lunar.fromYmd(y, month, day).getSolar();
        final date = DateTime(solar.getYear(), solar.getMonth(), solar.getDay());
        if (!date.isBefore(base)) return date;
      } catch (_) {
        // 该年没有这个农历日 (如 三十 在只有 29 天的月份) → 跳到下一年
        continue;
      }
    }
    return null;
  }

  // 阳历 (2/29 平年 → 按 3/1 算, 不会漏)
  DateTime build(int y) {
    final d = DateTime(y, month, day);
    if (d.month == month && d.day == day) return d;
    // 该年没有这个日期 (2/29) → 用 3/1
    return DateTime(y, 3, 1);
  }

  final thisYear = build(base.year);
  if (!thisYear.isBefore(base)) return thisYear;
  return build(base.year + 1);
}

/// 距下次生日还有几天 (含今天 = 0); 算不出来 null
int? daysUntilBirthday({
  required int? month,
  required int? day,
  String calendar = 'solar',
  DateTime? from,
}) {
  final next = nextBirthdayDate(
    month: month,
    day: day,
    calendar: calendar,
    from: from,
  );
  if (next == null) return null;
  final today = from ?? DateTime.now();
  final base = DateTime(today.year, today.month, today.day);
  return next.difference(base).inDays;
}

/// 完整信息 (文案 + 倒计时 + 下次阳历日期)
BirthdayInfo? birthdayInfo({
  required int? month,
  required int? day,
  String calendar = 'solar',
  int? year,
  DateTime? from,
}) {
  final next = nextBirthdayDate(
    month: month,
    day: day,
    calendar: calendar,
    from: from,
  );
  if (next == null) return null;
  final today = from ?? DateTime.now();
  final base = DateTime(today.year, today.month, today.day);
  return BirthdayInfo(
    daysUntil: next.difference(base).inDays,
    nextSolarDate: next,
    label: birthdayLabel(
      month: month,
      day: day,
      calendar: calendar,
      year: year,
    ),
  );
}

/// 是否落在「我设置的提醒窗口」内 (7/3/0 天前)
/// remindDays = null → 不提醒 → false
bool isInBirthdayRemindWindow({
  required int? month,
  required int? day,
  String calendar = 'solar',
  int? remindDays,
  DateTime? from,
}) {
  if (remindDays == null) return false;
  final d = daysUntilBirthday(
    month: month,
    day: day,
    calendar: calendar,
    from: from,
  );
  if (d == null) return false;
  return d <= remindDays;
}

/// 提醒强度文案
String remindLabel(int? remindDays) {
  switch (remindDays) {
    case 7:
      return '提前 7 天提醒';
    case 3:
      return '提前 3 天提醒';
    case 0:
      return '生日当天提醒';
    default:
      return '不提醒';
  }
}

File: src/lib/follow-up/birthday.ts
// ============================================
// 生日: 距下次生日还有几天 (服务端, 跟进紧急度用)
// ============================================
// ⚠️ 目前只支持**阳历** (solar); 农历生日服务端返回 null (不参与紧急度),
//    由客户端 (Flutter `core/utils/birthday.dart`, 有 lunar 包) 负责展示。
//    服务端要支持农历需要一个 TS 农历库 (lunar-typescript) 或预计算列 —— 见方案 §13 风险表。
//
// 边界 (与客户端一致):
//   - 月/日 任一为空 → null (只知道年份也算不出来)
//   - 2/29 在平年 → 按 2 月最后一天 (2/28) 算
//   - 只往前看 0-370 天

export interface BirthdayWindow {
  /** 距下次生日还有几天 (0 = 今天) */
  daysUntil: number;
  /** 提醒窗口 (7/3/0 天); null = 不提醒 */
  remindDays: number;
}

export function solarBirthdayWindow(
  month: number | null,
  day: number | null,
  calendar: string,
  remindDays: number | null,
  now: Date
): BirthdayWindow | null {
  if (month == null || day == null) return null;
  if (calendar !== "solar") return null; // 农历: 服务端暂不支持 (客户端展示)
  if (month < 1 || month > 12 || day < 1 || day > 31) return null;

  const today = new Date(
    Date.UTC(now.getUTCFullYear(), now.getUTCMonth(), now.getUTCDate())
  );

  const build = (year: number) => {
    // 2/29 平年 → 落到 2/28 (与客户端一致, 不报错)
    const lastDay = new Date(Date.UTC(year, month, 0)).getUTCDate();
    const d = Math.min(day, lastDay);
    return new Date(Date.UTC(year, month - 1, d));
  };

  let next = build(today.getUTCFullYear());
  if (next.getTime() < today.getTime()) {
    next = build(today.getUTCFullYear() + 1);
  }
  const daysUntil = Math.round((next.getTime() - today.getTime()) / 86_400_000);
  return { daysUntil, remindDays: remindDays ?? 0 };
}
